File: SpiritualCornerApp/src/types/prayer.ts
/**
 * Prayer Data Types
 * Defines the structure for the normalized prayer data schema.
 *
 * Schema: { categories: Category[], items: EventItem[] }
 * Navigation flow: Category -> EventItem -> Prayer
 */

// ─── New Normalized Schema Types ──────────────────────────────────

/** Top-level grouping (e.g. "VĂN KHẤN THẦN LINH TẠI GIA") */
export interface Category {
  id: string;
  name: string;
}

/** A single prayer/chant within an Event */
export interface Prayer {
  id: string;
  name: string;
  /** Optional per-prayer introduction or description */
  intro?: string;
  /** Full prayer/chant text (preserve \n line breaks) */
  content: string;
}

/** An event or ceremony containing one or more prayers */
export interface EventItem {
  id: string;
  /** References Category.id */
  categoryId: string;
  name: string;
  /** Always an array (can be empty). References lunar calendar date IDs. */
  prayDateIDs: string[];
  isDaily: boolean;
  isMonthly: boolean;
  /** Merged "Xuất Xứ" (origin/history) at Event level */
  detail?: string;
  /** Merged "Sắm Lễ" (offerings/preparations) at Event level */
  prepare?: string;
  /** Non-empty array of prayers */
  prayers: Prayer[];
}

/** Root data shape from the JSON file */
export interface PrayersData {
  categories: Category[];
  items: EventItem[];
}

// ─── Search ──────────────────────────────────────────────────────

export interface SearchResult {
  event: EventItem;
  matchingPrayerIds: string[];
  matchType: 'eventName' | 'prayerName' | 'prayerContent';
}

// ─── Legacy Types (deprecated – kept only for backward compat) ──

/** @deprecated Use EventItem instead */
export interface RawPrayerItem {
  id: string;
  number: string;
  title: string;
  content: string;
  isPrayer: boolean;
  prayDateID: string | string[] | null;
  children: RawPrayerItem[];
}

/** @deprecated Use EventItem instead */
export interface PrayerItem {
  id: string;
  number: string;
  title: string;
  content: string;
  isPrayer: boolean;
  prayDateID: string[];
  parentId: string | null;
  depth: number;
}

/** @deprecated Use Category instead */
export interface PrayerCategory {
  id: string;
  number: string;
  title: string;
  childCount: number;
}

/** @deprecated No longer used */
export interface PrayerPackage {
  id: string;
  number: string;
  title: string;
  items: PrayerItem[];
}

File: spiritual_app/lib/domain/models/prayer.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'prayer.freezed.dart';
part 'prayer.g.dart';

@freezed
class Prayer with _$Prayer {
  const factory Prayer({
    required String id,
    required String name,
    required String intro,
    required List<String> content,
  }) = _Prayer;

  factory Prayer.fromJson(Map<String, dynamic> json) => _$PrayerFromJson(json);
}

@freezed
class PrayerItem with _$PrayerItem {
  const factory PrayerItem({
    required String id,
    @JsonKey(name: 'categoryId') required String categoryId,
    required String name,
    required List<Prayer> prayers,
    List<String>? detail,
    List<String>? prepare,
  }) = _PrayerItem;

  factory PrayerItem.fromJson(Map<String, dynamic> json) => _$PrayerItemFromJson(json);
}

@freezed
class PrayerCategory with _$PrayerCategory {
  const factory PrayerCategory({
    required String id,
    required String name,
  }) = _PrayerCategory;

  factory PrayerCategory.fromJson(Map<String, dynamic> json) => _$PrayerCategoryFromJson(json);
}

@freezed
class PrayersData with _$PrayersData {
  const factory PrayersData({
    required List<PrayerCategory> categories,
    required List<PrayerItem> items,
  }) = _PrayersData;

  factory PrayersData.fromJson(Map<String, dynamic> json) =>
      _$PrayersDataFromJson(json);
}
